### ISSUE TITLE
LiGRU is 25 times slower than PyTorch GRU

### ISSUE THREAD

**Issue by sugeedarou** [speechbrain/speechbrain#793]
Hello,
maybe I'm using it incorrectly, but the LiGRU seems to be about 30 times slower than a normal PyTorch GRU which should not be the case.

![Screenshot from 2021-05-30 15-28-01](https://user-images.githubusercontent.com/6258568/120106083-edc18700-c15b-11eb-83fe-3e5e9ad93690.png)

Use case:
```import torch
import torch.nn as nn
import speechbrain as sb

from settings import *


class LiGRUModel(nn.Module):
    def __init__(self, output_size):
        super().__init__()
        self.output_size = output_size
        self.rnn = sb.nnet.RNN.LiGRU(input_shape=[64, 500, 89],
                           hidden_size=512,
                           num_layers=3,
                           nonlinearity='tanh',
                           normalization='None',
                           bidirectional=True)
        self.fc3 = nn.Linear(self.rnn.hidden_size*2, self.output_size)

    def forward(self, batch, lengths, device):
        out, _ = self.rnn(batch)
        out = self.fc3(out)
        predictions = torch.zeros(
            lengths.sum().item(), self.output_size, device=device)
        p = 0
        for i in range(batch.size(0)):
            predictions[p:p+lengths[i], :] = out[i][:lengths[i]]
            p += lengths[i]
        
        return predictions
```
All tensors and the model are on the GPU (additionally calling .cuda() on self.rnn does not make a difference)
**OS**: Ubuntu 20.04.2 LTS, **CPU**: Ryzen 7 3700x, **GPU**: GeForce RTX 2070 SUPER
The installation seems to be correct since pytest tests and pytest --doctest-modules speechbrain runs through without problems.

Thanks in advance!

**Comment by mravanelli** [comment]
Hi,
the LiGRU is a custom RNN and to make it faster you have to compile it with
JIT. This is what we do, for instance, here:
https://github.com/speechbrain/speechbrain/blob/develop/recipes/TIMIT/ASR/seq2seq/hparams/train.yaml#L179
The model compilation is done within the Brain class directly (
https://github.com/speechbrain/speechbrain/blob/develop/speechbrain/core.py#L1077
).
This should make the model 2/2.5x faster. However, it won't be as fast as a
native cudnn implementation like LSTM or GRU.
According to our benchmark on TIMIT the model is 25% slower than an LSTM
but it saves more parameters and memory.
The model will be faster than LSTM and GRU with a cudnn implementation.
The numbers you show, however, are quite far from that. Could you please
double check with our TIMIT benchmark? Just change the rnn_class
in  speechbrain.lobes.models.CRDNN.CRDNN.

**Comment by sugeedarou** [comment]
Thank you very much for you answer!
Actually I'm not using speechbrain directly but instead just trying to use only the LiGRU model in an existing PyTorch project.
I tried to compile it analog to the links you provided:
```
module = torch.jit.script(self.rnn._modules['rnn'])
self.rnn._modules['rnn'] = module.cuda()
```
But this unfortunately did not yield any performance improvement.
Could you please give me a short example of how I can compile it correctly?

**Comment by mravanelli** [comment]
My suggestion is to start from the TIMIT recipe in SpeechBrain. This is a
starting point that works for us. Make sure this is working properly on
your side as well. Once you check that this is fine, you can try to extract
step by step the model. It looks to me there is a bottleneck somewhere on
your side (e,g., very long sequences?), but it is easier to discover it if
you go step by step starting from a working implementation.

Keep us updated!

**Comment by sugeedarou** [comment]
I'm trying to use the LiGRU as an drop-in-replacement for a normal torch.nn.GRU which already trains really fast.
Maybe it is just not yet possible to use speechbrain in a performant way without the pipeline described in https://speechbrain.readthedocs.io/en/latest/experiment.html.
It would be really useful if these models could be used independent of the pipeline as such a comprehensive collection of state-of-the-art models is hard to find elsewhere.

**Comment by mravanelli** [comment]
This is already a stand-alone model that can be compiled. 
Please, compile it with: 
`rnn = torch.jit.script(rnn)`
This is what we do in https://github.com/speechbrain/speechbrain/blob/develop/speechbrain/core.py#L1077 and we don't observe issues.
Just to give you some numbers (minutes per-epoch) on our TIMIT benchmark:
```
LiGRU (not compiled): 4m 00s 
LiGRU (compiled): 2m 09s
GRU (cudnn): 1m 44 s
LSTM(cudnn): 1m 49 s
```
As you can see, the cudnn are still faster but now that much. If you see 25x different it looks like you have a bottleneck somewhere in your code. For instance, make sure you work with the latest version of pytorch (the pytorch team did several fixes in the recent versions).

**Comment by sugeedarou** [comment]
Thank you for your response!

Compiling works now, but the training time is still as slow as before.
If I find out the reason I will report it here.

**Comment by mravanelli** [comment]
That's weird. Which version of pytorch do you have? and how long are your
input sequences?

**Comment by mravanelli** [comment]
Any news of this?

**Comment by sugeedarou** [comment]
Unfortunately not, I was hoping someone else could confirm it.

I think it's important to mention, that it's being used inside a Pytorch Lightning project (called inside a pytorch_lightning.LightningModule and trained using pytorch_lightning.trainer), so maybe there is some interference here.

Thanks alot for your help so far!
For my use case the LiGRU did not give an improvement in accuracy so I'm keeping the regular GRU anyway.
